x-pack/plugins/siem/server/lib/hosts/elasticsearch_adapter.ts

```diff
@@ -4,67 +4,93 @@
       'search',
-      buildQuery(options)
+      buildHostsQuery(options)
     );
     const { cursor, limit } = options.pagination;
     const totalCount = getOr(0, 'aggregations.host_count.value', response);
-    const hits: HostHit[] = getOr([], 'aggregations.group_by_host.buckets', response).map(
-      (bucket: HostBucket) => ({
-        ...head(bucket.host.hits.hits),
-        cursor: bucket.key!.host_name,
-        firstSeen: bucket.firstSeen.value_as_string,
-      })
-    );
-    const hostsEdges = hits.map(hit => formatHostsData(options.fields, hit, hostsFieldsMap));
+    const buckets: HostAggEsItem[] = getOr([], 'aggregations.host_data.buckets', response);
+    const hostsEdges = buckets.map(bucket => formatHostEdgesData(options.fields, bucket));
     const hasNextPage = hostsEdges.length === limit + 1;
     const beginning = cursor != null ? parseInt(cursor, 10) : 0;
     const edges = hostsEdges.splice(beginning, limit - beginning);
 
+    return { edges, totalCount, pageInfo: { hasNextPage, endCursor: { value: String(limit) } } };
+  }
+
+  public async getHostDetails(
+    request: FrameworkRequest,
+    options: HostDetailsRequestOptions
+  ): Promise<HostItem> {
+    const response = await this.framework.callWithRequest<HostAggEsData, TermAggregation>(
+      request,
+      'search',
+      buildHostDetailsQuery(options)
+    );
+    const aggregations: HostAggEsItem = get('aggregations', response) || {};
+    return { _id: options.hostName, ...formatHostItem(options.fields, aggregations) };
+  }
+
+  public async getHostFirstLastSeen(
+    request: FrameworkRequest,
+    options: HostLastFirstSeenRequestOptions
+  ): Promise<FirstLastSeenHost> {
+    const response = await this.framework.callWithRequest<HostAggEsData, TermAggregation>(
+      request,
+      'search',
+      buildLastFirstSeenHostQuery(options)
+    );
+
+    const aggregations: HostAggEsItem = get('aggregations', response) || {};
     return {
-      edges,
-      totalCount,
-      pageInfo: {
-        hasNextPage,
-        endCursor: {
-          value: String(limit),
-        },
-      },
+      firstSeen: get('firstSeen.value_as_string', aggregations),
+      lastSeen: get('lastSeen.value_as_string', aggregations),
     };
   }
 }
 
-export const formatHostsData = (
+export const formatHostEdgesData = (
   fields: ReadonlyArray<string>,
-  hit: HostHit,
-  fieldMap: Readonly<Record<string, string>>
+  bucket: HostAggEsItem
 ): HostsEdges =>
   fields.reduce<HostsEdges>(
     (flattenedFields, fieldName) => {
-      flattenedFields.node._id = hit._id;
-      if (hit.cursor) {
-        flattenedFields.cursor.value = hit.cursor;
-      }
-      if (hit.firstSeen) {
-        flattenedFields.node.firstSeen = hit.firstSeen;
+      const hostId = get('key', bucket);
+      flattenedFields.node._id = hostId || null;
+      flattenedFields.cursor.value = hostId || '';
+
+      const fieldValue = getHostFieldValue(fieldName, bucket);
+      if (fieldValue) {
```

Comment:
You have to be careful about an `if` with a string like this because if this is an empty string then it will not evaluate as true, not just null values.

```ts
if ("") {
  console.log("You won't see this");
} else {
  console.log("You will see this");
}
You will see this
```
Is that your intention? If not, I would do a `if (fieldValue != null)` instead.